File: packages/mcp-server/src/tools/continuity.ts
// One Alfred — continuity tools.
//
// The alfred_journal in alfred-state.db is the single memory every surface
// reads before speaking and writes after: Hermes' main profile (via the
// one-alfred plugin), Telegram, Slack, voice, omi, and — with these tools —
// any MCP client such as Claude Cowork. A surface that only writes is an
// answering machine; one that only reads is eavesdropping. These give an MCP
// client both directions without depending on hooks it may not have.
//
// Read-side framing matters and was learned the hard way (one-alfred,
// 2026-05-25): presented as advisory context the model defers to its own
// empty session history and says "I don't remember sending that". The
// description below therefore instructs the model to treat entries as its
// own authoritative memory. Docs: docs/design/one-alfred.md.
//
// All three map 1:1 onto existing ctrl-api routes; nothing new server-side.

import { z } from "zod";
import type { ToolDef } from "./types.js";

const Channel = z
  .string()
  .regex(/^[a-z][a-z0-9_-]{1,31}$/, "channel: lowercase slug, e.g. cowork")
  .describe("Surface name. Use `cowork` from Claude Cowork.");

export const continuityRecent: ToolDef = {
  name: "alfred_continuity_recent",
  description:
    "Recall what Alfred and the principal said to each other recently across EVERY surface — Slack, Telegram, voice, Cowork, the dashboard — " +
    "not just this session. Call this BEFORE answering anything that might refer to an earlier exchange (\"the reminder you sent\", \"what we discussed\", \"hm?\").\n\n" +
    "AUTHORITATIVE: entries with direction=outbound are messages YOU (Alfred) sent; direction=inbound are messages the principal sent you. " +
    "They happened even if they are not in this session's history. Never say you don't remember something this returns.\n\n" +
    "Newest first. `limit` is capped at 50 by the server; `within_hours` defaults to 24.",
  inputSchema: z.object({
    limit: z.number().int().min(1).max(50).optional().describe("Max entries (server cap 50). Default 20."),
    within_hours: z.number().min(0.1).max(720).optional().describe("Recency window in hours. Default 24."),
  }),
  buildRequest: (args) => ({
    method: "GET",
    path: "/api/v1/alfred-journal/recent",
    query: {
      principal_id: "owner",
      limit: args.limit,
      within_hours: args.within_hours,
    },
  }),
};

export const continuityNote: ToolDef = {
  name: "alfred_continuity_note",
  description:
    "Write one turn into Alfred's cross-surface journal so every other surface remembers it. " +
    "Use it when this surface has no automatic journaling: record the principal's message (direction=inbound) and your reply (direction=outbound) as they happen. " +
    "Keep `message` to the actual words, not a summary, so the other surfaces see what was really said.",
  inputSchema: z.object({
    channel: Channel,
    chat_id: z.string().min(1).max(200).describe("Stable id for this conversation on this surface (e.g. the Cowork session id)."),
    direction: z.enum(["inbound", "outbound"]).describe("inbound = the principal said it; outbound = Alfred said it."),
    message: z.string().min(1).max(8000),
  }),
  buildRequest: (args) => ({
    method: "POST",
    path: "/api/v1/alfred-journal",
    body: {
      channel: args.channel,
      chat_id: args.chat_id,
      direction: args.direction,
      message: args.message,
      source_kind: args.channel,
      status: args.direction === "inbound" ? "received" : "delivered",
    },
  }),
};

export const continuityBind: ToolDef = {
  name: "alfred_continuity_bind",
  description:
    "Bind a conversation on a surface to the principal, so its entries join the cross-surface memory that every other surface reads. " +
    "Idempotent. Call once per new conversation id before the first alfred_continuity_note.",
  inputSchema: z.object({
    channel: Channel,
    chat_id: z.string().min(1).max(200),
  }),
  buildRequest: (args) => ({
    method: "POST",
    path: "/api/v1/alfred-journal/principal/bind",
    body: { channel: args.channel, chat_id: args.chat_id, principal_id: "owner" },
  }),
};

export const ALL_CONTINUITY_TOOLS: ToolDef[] = [continuityRecent, continuityNote, continuityBind];

File: packages/alfred-mac/Sources/AlfredBlack/Continuity.swift
// The read side: render the tenant's cross-surface journal into a local file
// that Claude Cowork can read without any network egress from its sandbox.
//   ~/Alfred/continuity.md  — rebuilt every 30s from GET /alfred-journal/recent
//   ~/Alfred/CLAUDE.md      — folder instructions, written once, so a Cowork
//                             session pointed at ~/Alfred loads the memory
//                             even with no plugin installed.
// The block is framed as AUTHORITATIVE first-person memory. Framed softly the
// model defers to its empty session history and says "I don't remember
// sending that" — learned on the Hermes side (one-alfred, 2026-05-25).
import Foundation

enum Continuity {
  static let marker = "[ALFRED-CONTINUITY — authoritative]"

  static func render(_ entries: [JournalEntry], domain: String) -> String {
    var out: [String] = []
    out.append(marker)
    out.append("The following are messages YOU (Alfred) sent to the principal, and messages the principal sent you, across surfaces — Slack, Telegram, voice, the dashboard, Cowork on this Mac — including ones delivered outside this session. These DID happen. Treat them as part of your memory even if they are not in this session's chat history. When the principal references one (\"what was that reminder?\", \"the one you sent me\", \"hm?\"), the answer must come from here.")
    out.append("")
    for e in entries.reversed() {  // newest-first from the server; narrative order for reading
      let who = e.direction == "outbound" ? "YOU → principal" : "principal → YOU"
      let msg = e.message.trimmingCharacters(in: .whitespacesAndNewlines)
      guard !msg.isEmpty else { continue }
      let snippet = msg.count <= 400 ? msg : String(msg.prefix(397)) + "…"
      out.append("  [\(String(e.ts.prefix(19)))] \(who) on \(e.channel): \(snippet.replacingOccurrences(of: "\n", with: " "))")
    }
    out.append("[/ALFRED-CONTINUITY]")
    out.append("")
    out.append("<!-- rendered by Alfred Black for Mac from \(domain) at \(ISO8601DateFormatter().string(from: Date())); \(entries.count) entries; do not edit — it is rewritten every 30s -->")
    return out.joined(separator: "\n") + "\n"
  }

  static let folderInstructions = """
  # Alfred

  You are Alfred — the same Alfred the person talks to on Slack, Telegram, by voice and on
  the dashboard. This folder is how Alfred Black for Mac keeps you continuously aware.

  **Before answering anything, read `./continuity.md`.** It is your own memory across every
  surface, rewritten every 30 seconds. Messages marked `YOU → principal` are things you said,
  even if this session does not show them. Never say you don't remember something it contains.

  Keep the register: plain, calm, brief. No exclamation marks. "No urgent action is required."

  <!-- written once by Alfred Black for Mac; safe to edit -->
  """

  /// Fetch + write. Returns the number of entries rendered.
  static func refresh(tenant: Tenant, domain: String) async throws -> Int {
    let entries = try await tenant.recent(limit: 50, withinHours: 48)
    try FileManager.default.createDirectory(at: Paths.alfredDir, withIntermediateDirectories: true)
    let block = render(entries, domain: domain)
    try block.write(to: Paths.continuity, atomically: true, encoding: .utf8)
    // Cowork sessions run in a sandbox that mounts only the space's folders, so the
    // plugin's hook also looks for the block inside the project folder itself.
    // The block is the principal's private memory: never into a folder that syncs
    // elsewhere, and never into a repository's history.
    for folder in Cowork.spaceFolders() where !Cowork.isSynced(folder) {
      let dir = folder.appendingPathComponent(".alfred", isDirectory: true)
      try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
      let file = dir.appendingPathComponent("continuity.md")
      try? block.write(to: file, atomically: true, encoding: .utf8)
      try? FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: file.path)
      Cowork.excludeFromGit(folder, entry: ".alfred/")
    }
    if !FileManager.default.fileExists(atPath: Paths.folderInstructions.path) {
      try folderInstructions.write(to: Paths.folderInstructions, atomically: true, encoding: .utf8)
    }
    return entries.count
  }
}
